sys/net/application_layer/gcoap/forward_proxy.c

```diff
@@ -0,0 +1,435 @@
+    if (pdu_len == -ENOMEM) {
+        return gcoap_response(pdu, buf, len, COAP_CODE_INTERNAL_SERVER_ERROR);
+    }
+    /* Proxy-Uri malformed, reply with 4.02 */
+    else if (pdu_len == -EINVAL) {
+        return gcoap_response(pdu, buf, len, COAP_CODE_BAD_OPTION);
+    }
+    /* scheme not supported */
+    else if (pdu_len == -EPERM) {
+        return gcoap_response(pdu, buf, len, COAP_CODE_PROXYING_NOT_SUPPORTED);
+    }
+
+    return pdu_len;
+}
+
+static bool _parse_endpoint(sock_udp_ep_t *remote,
+                            uri_parser_result_t *urip)
+{
+    char scratch[8];
+    ipv6_addr_t addr;
+    remote->family = AF_INET6;
+
+    /* support IPv6 only for now */
+    if (!urip->ipv6addr) {
+        return false;
+    }
+
+    /* check for interface */
+    if (urip->zoneid) {
+        /* only works with integer based zoneids */
+
+        if (urip->zoneid_len > (sizeof(scratch)/sizeof(scratch[0]) - 1)) {
+            return false;
+        }
+
+        memcpy(scratch, urip->zoneid, urip->zoneid_len);
+
+        scratch[urip->zoneid_len] = '\0';
+
+        int pid = atoi(scratch);
+
+        if (gnrc_netif_get_by_pid(pid) == NULL) {
+            return false;
+        }
+        remote->netif = pid;
+    }
+    /* no interface present */
+    else {
+        if (gnrc_netif_numof() == 1) {
+            /* assign the single interface found in gnrc_netif_numof() */
+            remote->netif = (uint16_t)gnrc_netif_iter(NULL)->pid;
+        }
+        else {
+            remote->netif = SOCK_ADDR_ANY_NETIF;
+        }
+    }
+
+    /* parse destination address */
+    if (ipv6_addr_from_buf(&addr, urip->ipv6addr, urip->ipv6addr_len) == NULL) {
+        return false;
+    }
+    if ((remote->netif == SOCK_ADDR_ANY_NETIF) &&
+        ipv6_addr_is_link_local(&addr)) {
+        return false;
+    }
+    memcpy(&remote->addr.ipv6[0], &addr.u8[0], sizeof(addr.u8));
+
+    if (urip->port_len) {
+        /* copy port string into scratch for atoi */
+        memcpy(scratch, urip->port, urip->port_len);
+        scratch[urip->port_len] = '\0';
+
+        remote->port = atoi(scratch);
+
+        if (remote->port == 0) {
+          return false;
+        }
+    }
+    else {
+        remote->port = 5683;
+    }
+
+    return true;
+}
+
+static void _forward_resp_handler(const gcoap_request_memo_t *memo,
+                                  coap_pkt_t* pdu,
+                                  const sock_udp_ep_t *remote)
+{
+    (void) remote; /* this is the origin server */
+    client_ep_t *cep = (client_ep_t *)memo->context;
+
+    /* forward the response packet as-is to the client */
+    gcoap_forward_proxy_dispatch((uint8_t *)pdu->hdr,
+                                 (pdu->payload -
+                                  (uint8_t *)pdu->hdr + pdu->payload_len),
+                                 &cep->ep);
+
+#if IS_ACTIVE(MODULE_NANOCOAP_CACHE)
+    coap_pkt_t req;
+    req.hdr = (coap_hdr_t *) &memo->msg.hdr_buf[0];
```

Comment:
Not sure what is in this buffer, but in my experiments my FETCH request starts with `42 05 0c 7a`, but whatever is stored here starts with `48 03 00 20`, so the cache is missed, since the proxy thinks the request was a PUT request... :-/ Not really sure where this data is coming from. I don't see that sequence on the air.

Reply:
Ah, `msg` is a union of different types, depending on if the request was a NON or a CON. When it is a CON, it is a `gcoap_resend_t` struct, which has a pointer to the request as a first member. That is what I am reading. The correct way to get the request header is

```suggestion
    if (memo->send_limit == GCOAP_SEND_LIMIT_NON) {
        req.hdr = (coap_hdr_t *) &memo->msg.hdr_buf[0];
    }
    else {
        req.hdr = (coap_hdr_t *) memo->msg.data.pdu_buf;
    }
```

Reply:
Nice catch!  Will amend the proposed fix accordingly :+1:

Reply:
Note to self: maybe some `get_req_hdr(memo)` function can prevent such pitfalls in the future.